examples/pytorch/llm_agent/_common.py

```diff
@@ -0,0 +1,366 @@
+                 tokenize_function) -> None:
+        self._data = []
+        for i in tqdm(range(len(system))):
+            _d = tokenize_function(system[i], user[i], assistant[i])
+            if len(_d) == 0:
+                continue
+            self._data.append(_d)
+
+    def __getitem__(self, idx: int) -> Dict[str, Any]:
+        return self._data[idx]
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+def stat_dataset(dataset: "MyDataset") -> None:
+    """对数据集进行统计分析"""
+    _token_len = []
+    for d in dataset:
+        _token_len.append(len(d["input_ids"]))
+    _token_len = np.array(_token_len)
+    mean = _token_len.mean().item()
+    std = _token_len.std().item()
+    min_ = _token_len.min().item()
+    max_ = _token_len.max().item()
+    logger.info(f"Dataset Token Length: {mean:.6f}±{std:.6f}, min={min_:.6f}, max={max_:.6f}, size={_token_len.shape[0]}")
+
+def print_examples(examples: Dict[str, Any], tokenizer) -> None:
+    input_ids, labels = examples["input_ids"], examples["labels"]
+    print(f"[INPUT_IDS] {tokenizer.decode(input_ids)}")
+    print()
+    print(f"[LABLES] {tokenizer.decode([l if l != -100 else 0 for l in labels])}")
+
+# 
+def data_collate_fn(batch: List[Dict[str, Any]], tokenizer) -> Dict[str, Any]:
+    input_ids = [torch.tensor(b["input_ids"]) for b in batch]
+    labels = [torch.tensor(b["labels"]) for b in batch]
+    attention_mask = [torch.ones(len(input_ids[i]), dtype=torch.int64) for i in range(len(input_ids))]
+    #
+    input_ids = pad_sequence(input_ids, batch_first=True, padding_value=tokenizer.pad_token_id)
+    attention_mask = pad_sequence(attention_mask, batch_first=True, padding_value=0)
+    labels = pad_sequence(labels, batch_first=True, padding_value=-100)
+    return {"input_ids": input_ids, "attention_mask": attention_mask, "labels": labels}
+
+
+def print_model_info(model: Module, name: Optional[str] = None) -> None:
+    if name is None:
+        name = model.__class__.__name__
+    #
+    n_params = sum(p.numel() for p in model.parameters())
+    n_grads = sum(p.numel() for p in model.parameters() if p.requires_grad)
+    n_buffers = sum(p.numel() for p in model.buffers())
+    #
+    n_params /= 1e6
+    n_grads /= 1e6
+    n_buffers /= 1e6
+    s = [
+        f"{name}: ",
+        f"{n_params:.4f}M Params ({n_grads:.4f}M Trainable), ",
+        f"{n_buffers:.4f}M Buffers",
+    ]
+    s += "."
+    logger.info("".join(s))
+
+def show_freeze_layers(model: Module, max_lines: int = 20) -> None:
+    named_p = list(model.named_parameters())
+    for i, (n, p) in enumerate(named_p):
+        if i >= max_lines:
+            logger.info("...")
+            break
+        logger.info(f"{n}: requires_grad={p.requires_grad}")
+
+@METRICS.register_module(group_key=default_group, module_name='my_metric')
+class MyMetric(Metric):
+    def __init__(self, vocab_size: int):
+        self.acc = Accuracy("multiclass", num_classes=vocab_size)
```

Comment:
Why acc, not Rouge or Bleu?

Reply:
The rouge bleu based approach requires a token-by-token generation approach, which leads to long validation times. 
acc is used here for simple validation.